Sources/Internal/GrpcClient.swift

```diff
@@ -151,25 +151,49 @@ actor GrpcClient: CustomStringConvertible {
       DataConnectLogger
         .debug("executeQuery() receives response: \(resultsString, privacy: .private).")
 
-      // Not doing error decoding here
-      guard results.errors.isEmpty else {
-        throw DataConnectError
-          .operationExecutionFailed(messages: createErrorJson(errors: results.errors))
-      }
-
-      if let decodedResults = try codec.decode(result: results.data, asType: resultType) {
-        return OperationResult(data: decodedResults)
-      } else {
-        // In future, set this as error in OperationResult
-        DataConnectLogger
-          .debug("executeQuery() response: \(resultsString, privacy: .private) decode failed.")
-        throw DataConnectError.decodeFailed
+      // lets decode partial errors. We need these whether we succeed or fail
+      let errorInfoList = createErrorInfoList(errors: results.errors)
+
+      // check if decode succeeds
+      do {
+        let decodedResults = try codec.decode(result: results.data, asType: resultType)
+
+        // even though decode succeeded, we may still have received partial errors
+        if !errorInfoList.isEmpty {
+          let failureResponse = OperationFailureResponse(
+            rawJsonData: resultsString, errors: errorInfoList,
+            data: decodedResults
+          )
+          throw DataConnectOperationError.executionFailed(
+            response: failureResponse
+          )
+        } else {
+          return OperationResult(data: decodedResults)
+        }
+
+      } catch {
+        // we failed to decode
+        if !errorInfoList.isEmpty {
+          let failureResponse = OperationFailureResponse(
+            rawJsonData: resultsString,
+            errors: errorInfoList,
+            data: nil
+          )
+          throw DataConnectOperationError
+            .executionFailed(
+              cause: error,
+              response: failureResponse
+            )
```

Comment:
IIUC: an error with be throw if `!errorInfoList.isEmpty == true` on line 167, that will be quickly re-caught in the highlighted catch block where `!errorInfoList.isEmpty` will be `true` again and we will unnecessarily? re-wrap the thrown error.

Reply:
Thanks for catching this. The logic is a bit awkward. I've fixed this and also added a note about the logic. The inner error throws with decodedData while outer throws without decodedData.